cmd/generate-catalog/generate.go

```diff
@@ -0,0 +1,227 @@
+	var channel *Channel
+
+	for _, v := range versions {
+		// Refresh major channel entries list when new major version is reached
+		if v.Major() != previousEntryVersion.Major() {
+			majorEntries = make([]ChannelEntry, 0)
+		}
+
+		// Create a new channel entry for each new minor version (patch = 0)
+		if v.Patch() == 0 {
+			previousChannelVersion = previousEntryVersion
+			if channel != nil {
+				channels = append(channels, *channel)
+				// Add "latest" channel when all "3.X.X" version are processed
+				if v.Original() == "4.0.0" {
+					latestChannel := newLatestChannel(channel.Entries)
+					channels = append(channels, latestChannel)
+				}
+			}
+			channel = newChannel(v, slices.Clone(majorEntries))
+		}
+
+		catalogChannelEntry := newChannelEntry(v, previousEntryVersion, previousChannelVersion, brokenVersions)
+		channel.Entries = append(channel.Entries, catalogChannelEntry)
+		majorEntries = append(majorEntries, catalogChannelEntry)
+
+		previousEntryVersion = v
+	}
+
+	// Add the last channel entry for the last version processed
+	channels = append(channels, *channel)
+	// Add "stable" channel when the last version is reached
+	stableChannel := newStableChannel(channel.Entries)
+	channels = append(channels, stableChannel)
+
+	return channels
+}
+
+// generateDeprecations creates an object with a list of deprecations based on the provided versions.
+func generateDeprecations(versions []*semver.Version, oldestSupportedVersion *semver.Version) Deprecations {
+	var deprecations []DeprecationEntry
+	var channelVersions []*semver.Version
+	for _, v := range versions {
+		// Each 0 Patch version indicates a new channel.
+		if v.Patch() == 0 {
+			channelVersions = append(channelVersions, v)
+		}
+	}
+
+	// Deprecate all channels that are older than the oldest supported version
+	for _, channelVersion := range channelVersions {
+		if channelVersion.LessThan(oldestSupportedVersion) {
+			deprecations = append(deprecations, newChannelDeprecationEntry(channelVersion))
+		}
+	}
+
+	// Deprecate all bundles that are older than the oldest supported version
+	for _, v := range versions {
+		if v.LessThan(oldestSupportedVersion) {
+			deprecations = append(deprecations, newBundleDeprecationEntry(v))
+		}
+	}
+
+	return newDeprecation(deprecations)
+}
+
+// generateBundles creates a list of bundle entries based on the provided versions and their corresponding images.
+func generateBundles(versions []*semver.Version, versionToImageMap map[*semver.Version]BundleImage) []BundleEntry {
+	var bundleEntries []BundleEntry
+	for _, v := range versions {
+		bundleEntries = append(bundleEntries, newBundleEntry(versionToImageMap[v].Image))
+	}
+	return bundleEntries
+}
+
+func generateChannelName(version *semver.Version) string {
+	return fmt.Sprintf("rhacs-%d.%d", version.Major(), version.Minor())
+}
+
+func generateBundleName(version *semver.Version) string {
+	return fmt.Sprintf("rhacs-operator.v%d.%d.%d", version.Major(), version.Minor(), version.Patch())
```

Comment:
It should rather use `version.Original()` to allow for versions like `4.9.0-281-g0a61145c18-fast`. See [my other comment](https://github.com/stackrox/operator-index/pull/137#discussion_r2254937116). Untested:
```suggestion
	return fmt.Sprintf("rhacs-operator.v%s", version.Original())
```